### PR TITLE
Third party auth providers

### PR DESCRIPTION
**Pull request by sdimitris**
Made OAuth2 strategy for authenticating with third party providers.

Implemented providers:
* Facebook
* Google
* Slack
* Figma
* Github
* Microsoft
* Twitch

### CODE REVIEW

File: modules/authentication/src/config/config.ts [hunk 1]
```diff
@@ -47,6 +47,16 @@ export default {
       default: '',
       nullable: true,
     },
+    clientSecret: {
+      format: 'String',
+      default: '',
+      optional: true,
+    },
+    redirect_uri: {
```
Comment (kon14) [inline modules/authentication/src/config/config.ts:55]: Fix camelCase

File: modules/authentication/src/config/config.ts [hunk 2]
```diff
@@ -64,6 +74,16 @@ export default {
       format: 'String',
       default: '',
     },
+    clientSecret: {
+      format: 'String',
+      default: '',
+      optional: true,
+    },
+    redirect_uri: {
```
Comment (kon14) [inline modules/authentication/src/config/config.ts:82]: Fix camelCase

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import { Payload } from '../interfaces/Payload';
+import { isNil } from 'lodash';
+import { status } from '@grpc/grpc-js';
+import { User } from '../../models';
+import { AuthUtils } from '../../utils/auth';
+import { ConfigController } from '../../config/Config.controller';
+import { OIDCSettings } from '../interfaces/OIDCSettings';
+import axios, { AxiosRequestConfig } from 'axios';
+
+export abstract class AuthenticationProviderClass<T extends Payload> {
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:15]: Drop "Class" suffix? AuthenticationProviderClass -> AuthenticationProvider

File: modules/authentication/src/handlers/facebook.ts [hunk 4]
```diff
@@ -32,76 +29,32 @@ export class FacebookHandlers {
...
 
-    const facebookResponse = await axios(facebookOptions);
-
+    const facebookResponse:any = await axios(facebookOptions);
     if (isNil(facebookResponse.data.email) || isNil(facebookResponse.data.id)) {
       throw new GrpcError(status.UNAUTHENTICATED, 'Authentication with facebook failed');
     }
-
-    let user: User | null = await User.getInstance().findOne({
+    let payload: Payload = {
```
Comment (kon14) [inline modules/authentication/src/handlers/facebook.ts:52]: return object directly (sans assignment) or at least make payload const.

File: modules/authentication/src/handlers/facebook.ts [hunk 5]
```diff
@@ -32,76 +29,32 @@ export class FacebookHandlers {
...
     if (!this.initialized)
       throw new GrpcError(status.NOT_FOUND, 'Requested resource not found');
-    const { access_token } = call.request.params;
-
-    const config = ConfigController.getInstance().config;
-
+    let { access_token } = call.request.params;
     const context = call.request.context;
-    if (isNil(context) || isEmpty(context))
+    if (( isNil(context) || isEmpty(context)) && !call.request.path.startsWith('/hook'))
```
Comment (kon14) [inline modules/authentication/src/handlers/facebook.ts:37]: spacing

File: modules/authentication/src/handlers/facebook.ts [hunk 6]
```diff
@@ -32,76 +29,32 @@ export class FacebookHandlers {
...
 
-  async authenticate(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+  async connectWithProvider(call: ParsedRouterRequest) {
     if (!this.initialized)
       throw new GrpcError(status.NOT_FOUND, 'Requested resource not found');
-    const { access_token } = call.request.params;
-
-    const config = ConfigController.getInstance().config;
-
+    let { access_token } = call.request.params;
```
Comment (kon14) [inline modules/authentication/src/handlers/facebook.ts:35]: const was fine

File: modules/authentication/src/handlers/google.ts [hunk 7]
```diff
@@ -1,22 +1,17 @@
...
+import { status } from '@grpc/grpc-js';
 
-export class GoogleHandlers {
+export class GoogleHandlers extends AuthenticationProviderClass<Payload> {
   private readonly client: OAuth2Client;
   private initialized: boolean = false;
 
-  constructor(private readonly grpcSdk: ConduitGrpcSdk) {
+  constructor(grpcSdk: ConduitGrpcSdk, url: string) {
+    super(grpcSdk,'google', url);
```
Comment (kon14) [inline modules/authentication/src/handlers/google.ts:14]: spacing

File: modules/authentication/src/handlers/google.ts [hunk 8]
```diff
@@ -35,7 +30,7 @@ export class GoogleHandlers {
     return true;
   }
 
-  async authenticate(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+  async connectWithProvider(call: ParsedRouterRequest): Promise<any> {
```
Comment (kon14) [inline modules/authentication/src/handlers/google.ts:33]: return type is `Promise<Payload>`

File: modules/authentication/src/handlers/google.ts [hunk 9]
```diff
@@ -58,56 +53,18 @@ export class GoogleHandlers {
...
-      user = await User.getInstance().create({
-        email: payload.email,
-        google: {
-          id: payload.sub,
-          token: access_token,
-          tokenExpires: moment().add(expires_in).format(),
-        },
-        isVerified: true,
-      });
+    let googlePayload: Payload = {
```
Comment (kon14) [inline modules/authentication/src/handlers/google.ts:61]: return object directly (sans assignment) or at least make googlePayload const.

File: modules/authentication/src/handlers/interfaces/Payload.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export interface Payload {
+  id: string;
+  email?:string;
```
Comment (kon14) [inline modules/authentication/src/handlers/interfaces/Payload.ts:3]: spacing

File: modules/authentication/src/handlers/interfaces/Payload.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+  clientId: string;
+  [field: string]:any;
```
Comment (kon14) [inline modules/authentication/src/handlers/interfaces/Payload.ts:5]: spacing

File: modules/authentication/src/routes/routes.ts [hunk 12]
```diff
@@ -28,11 +30,12 @@ export class AuthenticationRoutes {
   private readonly serviceHandler: ServiceHandler;
   private readonly commonHandlers: CommonHandlers;
   private readonly twitchHandlers: TwitchHandlers;
+  private options: OIDCSettings;
 
   constructor(readonly server: GrpcServer, private readonly grpcSdk: ConduitGrpcSdk) {
     this.localHandlers = new LocalHandlers(grpcSdk);
-    this.facebookHandlers = new FacebookHandlers(grpcSdk);
-    this.googleHandlers = new GoogleHandlers(grpcSdk);
+    this.facebookHandlers = new FacebookHandlers(grpcSdk,"https://graph.facebook.com/v12.0/oauth/access_token?");
```
Comment (kon14) [inline modules/authentication/src/routes/routes.ts:37]: spacing

File: modules/authentication/src/routes/routes.ts [hunk 13]
```diff
@@ -48,20 +51,23 @@ export class AuthenticationRoutes {
...
+        beginAuthFacebook: this.facebookHandlers.beginAuth(this.grpcSdk,this.options).bind(this.facebookHandlers),
+        authenticateFacebook: this.facebookHandlers.authenticate.bind(this.facebookHandlers),
+
+        beginAuthGoogle: this.googleHandlers.beginAuth(this.grpcSdk,this.options).bind(this.googleHandlers),
         authenticateGoogle: this.googleHandlers.authenticate.bind(this.googleHandlers),
+
         authenticateService: this.serviceHandler.authenticate.bind(this.serviceHandler),
         authenticateTwitch: this.twitchHandlers.authenticate.bind(this.twitchHandlers),
-        beginAuthTwitch: this.twitchHandlers.beginAuth.bind(this.twitchHandlers),
+        beginAuthTwitch: this.twitchHandlers.beginAuth(this.grpcSdk,this.options).bind(this.twitchHandlers),
```
Comment (kon14) [inline modules/authentication/src/routes/routes.ts:70]: spacing

File: modules/authentication/src/routes/routes.ts [hunk 14]
```diff
@@ -335,8 +393,38 @@ export class AuthenticationRoutes {
...
+          {
+            path: '/hook/google',
+            action: ConduitRouteActions.GET,
+            description: `Login/register with Google using redirection mechanism.`,
+            urlParams: {
+              code: ConduitString.Required,
+              state: ConduitString.Required,
+            },
+          },
+          new ConduitRouteReturnDefinition('FacebookResponse', {
```
Comment (kon14) [inline modules/authentication/src/routes/routes.ts:410]: This is Google.

File: modules/authentication/src/handlers/twitch.ts [hunk 15]
```diff
@@ -84,63 +68,14 @@ export class TwitchHandlers {
 
     id = response2.data.data[0].id;
     email = response2.data.data[0].email;
-    profile_image_url = response2.data.data[0].profile_image_url;
+    //profile_image_url = response2.data.data[0].profile_image_url;
```
Comment (kon14) [inline modules/authentication/src/handlers/twitch.ts:71]: ?

File: modules/authentication/src/handlers/twitch.ts [hunk 16]
```diff
@@ -30,24 +24,15 @@ export class TwitchHandlers {
...
-    const config = ConfigController.getInstance().config;
-
-    const serverConfig = await this.grpcSdk.config.getServerConfig();
-    const redirect = serverConfig.url + '/hook/authentication/twitch';
-    const clientId = call.request.context.clientId;
-    return `https://id.twitch.tv/oauth2/authorize?client_id=${config.twitch.clientId}&redirect_uri=${redirect}&response_type=code&scope=user:read:email&state=${clientId}`;
-  }
-
-  async authenticate(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+  async connectWithProvider(call:ParsedRouterRequest ): Promise<any> {
```
Comment (kon14) [inline modules/authentication/src/handlers/twitch.ts:35]: spacing

File: modules/authentication/src/handlers/twitch.ts [hunk 17]
```diff
@@ -1,21 +1,15 @@
...
-import moment from 'moment';
 import { ConfigController } from '../config/Config.controller';
-import { AuthUtils } from '../utils/auth';
-import { User } from '../models';
-import { status } from '@grpc/grpc-js';
+import { AuthenticationProviderClass } from './models/AuthenticationProviderClass';
+import { Payload } from './interfaces/Payload';
 
-export class TwitchHandlers {
+export class TwitchHandlers extends AuthenticationProviderClass<any> {
```
Comment (kon14) [inline modules/authentication/src/handlers/twitch.ts:7]: Use T where `<T extends Payload>`

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    const [accessToken, refreshToken] = await AuthUtils.createUserTokensAsPromise(
+      this.grpcSdk,
+      {
+        userId: user!._id,
+        clientId: payload.clientId,
+        config: config,
+      },
+    );
+
+    let retObject: any = {
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:119]: make const

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+  grpcSdk: ConduitGrpcSdk;
+
+  constructor(grpcSdk: ConduitGrpcSdk, providerName: string, url: string) {
+    this.providerName = providerName;
+    this.grpcSdk = grpcSdk;
+    this.accessTokenUrl = url;
+  }
+
+  abstract async validate(): Promise<Boolean>;
+  abstract async connectWithProvider(call: ParsedRouterRequest): Promise<any>;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:27]: Return type is `Promise<Payload>`

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+      url: this.accessTokenUrl,
+      params: {
+        client_id: config[this.providerName].clientId,
+        client_secret: config[this.providerName].clientSecret,
+        code: params.code,
+        redirect_uri: url + call.request.path,
+      },
+    };
+    const providerResponse: any = await axios(providerOptions);
+    let access_token = providerResponse.data.access_token;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:46]: make these two const. I guess the snake_case is used for compat with sth?

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  abstract async validate(): Promise<Boolean>;
+  abstract async connectWithProvider(call: ParsedRouterRequest): Promise<any>;
+
+
+  async redirect(call: ParsedRouterRequest) {
+    const params = call.request.params;
+    const config = ConfigController.getInstance().config;
+    let serverConfig = await this.grpcSdk.config.getServerConfig();
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:33]: make these two const.

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+    };
+    const providerResponse: any = await axios(providerOptions);
+    let access_token = providerResponse.data.access_token;
+    let clientId = params.state;
+    return { access_token, clientId };
+  }
+
+  beginAuth(grpcSdk:  ConduitGrpcSdk,oidcSettings: OIDCSettings) {
+    return async function(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+      let options = oidcSettings.getOptions;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:53]: const options and length

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+    let payload: any;
+    const config = ConfigController.getInstance().config;
+    if (call.request.path.startsWith('/hook')) {
+      payload = await this.redirect(call);
+      call.request.params['access_token'] = payload.access_token;
+      call.request.context['clientId'] = payload.clientId;
+    }
+    payload = await this.connectWithProvider(call);
+
+    let user: User | null | any = null;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:77]: why 'any'?

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+      if (isNil(user[this.providerName])) {
+        user[this.providerName] = payload;
+        // TODO look into this again, as the email the user has registered will still not be verified
+        if (!user.isVerified) user.isVerified = true;
+        user = await User.getInstance().findByIdAndUpdate(user._id, user);
+      }
+    } else {
+      user = await User.getInstance().create({
+        email: payload.email,
+        [this.providerName]:  payload,
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:106]: spacing

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    let clientId = params.state;
+    return { access_token, clientId };
+  }
+
+  beginAuth(grpcSdk:  ConduitGrpcSdk,oidcSettings: OIDCSettings) {
+    return async function(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+      let options = oidcSettings.getOptions;
+      let retUrl = options.url;
+      let length = Object.keys(options).length;
+     //if (options.hasOwnProperty('state')) options['state'] = call.request.context.clientId;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:56]: ?

File: modules/authentication/src/handlers/interfaces/OIDCSettings.ts [hunk 26]
```diff
@@ -0,0 +1,10 @@
+export class OIDCSettings {
+    private options: any;
+
+    constructor(options: any,state: string = 'state') {
+        options['state'] = state;
+        this.options = options;
+    }
+
+    get getOptions() { return  this.options; }
```
Comment (kon14) [inline modules/authentication/src/handlers/interfaces/OIDCSettings.ts:9]: spacing

File: modules/authentication/src/handlers/models/AuthenticationProviderClass.ts [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+          retUrl += k + '=' + options[k]
+          if ( i !== length -1) retUrl += '&'
+        }
+      });
+      return retUrl;
+    }
+  }
+
+  async authenticate(call: ParsedRouterRequest): Promise<UnparsedRouterResponse> {
+    let payload: any;
```
Comment (kon14) [inline modules/authentication/src/handlers/models/AuthenticationProviderClass.ts:68]: this is { access_token, string, clientId: string }

**Review by kon14 — CHANGES_REQUESTED** [review]

**Review by kkopanidis — CHANGES_REQUESTED** [review]
fix camelCase naming in variables
add default settings for providers
remove settings interfaces since they are not needed

**Review by kkopanidis — APPROVED** [review]

**Pull request merged**